### PR TITLE
import pkg:intl when DateFormat or NumberFormat is used

### LINKED ISSUE #82713: gen_l10n: Using DateTime produces error during code generation
Using `DataTime` in an `app_en.arb` file produces an error during code generation.

Example `app_en.arb` that reproduces the error:

```json
{
  "helloWorldOn": "Hello World on {date}",
  "@helloWorldOn": {
    "description": "A message with a date parameter",
    "placeholders": {
      "date": {
        "type": "DateTime",
        "format": "yMd"
      }
    }
  }
}
```

Error shown:

```
.dart_tool/flutter_gen/gen_l10n/app_localizations_en.dart:12:11: Error: 'intl.DateFormat' can't be used as a type because 'intl' doesn't refer to an import prefix.
    final intl.DateFormat dateDateFormat = intl.DateFormat.yMd(localeName);
          ^^^^^^^^^^^^^^^
.dart_tool/flutter_gen/gen_l10n/app_localizations_en.dart:12:44: Error: The getter 'intl' isn't defined for the class 'AppLocalizationsEn'.
 - 'AppLocalizationsEn' is from 'package:flutter_gen/gen_l10n/app_localizations_en.dart' ('.dart_tool/flutter_gen/gen_l10n/app_localizations_en.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'intl'.
    final intl.DateFormat dateDateFormat = intl.DateFormat.yMd(localeName);

```

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/82713.

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/generate_localizations_test.dart [hunk 1]
```diff
@@ -1489,6 +1489,45 @@ import 'output-localization-file_en.dart' deferred as output-localization-file_e
...
+            fileSystem: fs,
+            inputPathString: defaultL10nPathString,
+            outputPathString: defaultL10nPathString,
+            templateArbFileName: defaultTemplateArbFileName,
+            outputFileString: defaultOutputFileString,
+            classNameString: defaultClassNameString,
+          )
+            ..loadResources()
+            ..writeOutputFiles(BufferLogger.test());
+        } on Exception catch (e) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/generate_localizations_test.dart:1521]: similar question here, shouldn't this fail if already if an exception is thrown?

File: packages/flutter_tools/test/general.shard/generate_localizations_test.dart [hunk 2]
```diff
@@ -1565,6 +1604,48 @@ import 'output-localization-file_en.dart' deferred as output-localization-file_e
...
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/generate_localizations_test.dart:1640]: and here

File: packages/flutter_tools/test/general.shard/generate_localizations_test.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter_tools/test/general.shard/generate_localizations_test.dart:1521]: I'll remove it, was just copying from the other test cases.

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 3]
```diff
@@ -1168,7 +1168,7 @@ class LocalizationsGenerator {
       .replaceAll('\n\n\n', '\n\n');
   }
 
-  bool _containsPluralMessage() => _allMessages.any((Message message) => message.isPlural);
+  bool _requiresIntlImport() => _allMessages.any((Message message) => message.isPlural || message.placeholdersRequireFormatting);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1171]: is the plan to keep adding onto this or is there a comprehensive list of intl type usage?

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit + question

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1171]: As far as I can tell, it's a case by case decision whether a feature needs intl or not. If we add more features, this list would grow.

**Pull request merged**